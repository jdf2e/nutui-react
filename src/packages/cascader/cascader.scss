@import '../popup/popup.scss';
@import '../tabs/tabs.scss';

.nut-cascader {
  width: 100%;
  font-size: $cascader-font-size;

  .nut-tabs-titles {
    padding: $cascader-tabs-item-padding;
    background: $color-background-overlay;
  }

  .nut-tabs-titles-item {
    flex: initial;
    min-width: auto;
    width: auto;
    padding: $cascader-tabs-item-padding;
    white-space: nowrap;
  }

  .nut-tabpane {
    padding: 0;
    background: $color-background-overlay;
  }

  &-pane {
    display: block;
    width: 100%;
    padding-top: $cascader-pane-paddingTop;
    height: $cascader-pane-height;
    overflow-y: auto;
    -webkit-overflow-scrolling: touch;
  }

  &-item {
    display: flex;
    align-items: center;
    justify-content: center;
    // height: $cascader-item-height;
    padding: $cascader-item-padding;
    margin: $cascader-item-margin;
    border-bottom: $cascader-item-border-bottom;
    font-size: $cascader-item-font-size;
    color: $cascader-item-color;
    cursor: pointer;

    &.disabled {
      opacity: 0.6;
      cursor: not-allowed;
    }

    &.active {
      &:not(.disabled) {
        color: $cascader-item-active-color;
      }

      .nut-cascader-item-icon-check {
        visibility: visible;
        color: inherit;
      }
    }
  }

  &-item-title {
    flex: 1;
  }

  .nut-icon-checklist {
    margin-left: $cascader-icon-checklist-marginLeft;
    visibility: hidden;
  }
}

[dir='rtl'] .nut-cascader {
  .nut-icon-checklist {
    margin-left: 0;
    margin-right: $cascader-icon-checklist-marginLeft;
  }
}
