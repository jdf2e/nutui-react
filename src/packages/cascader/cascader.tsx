import React, {
  ForwardRefRenderFunction,
  PropsWithChildren,
  useState,
  useEffect,
} from 'react'
import classNames from 'classnames'
import Popup from '@/packages/popup'
import { Tabs } from '@/packages/tabs/tabs'
import { CascaderItem } from './cascaderItem'
import { convertListToOptions } from './helper'
import {
  CascaderPane,
  CascaderOption,
  CascaderValue,
  optionKey,
  format,
} from './types'
import Tree from './tree'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'

export interface CascaderProps extends BasicComponent {
  popup: boolean
  visible: boolean // popup 显示状态
  activeColor: string
  activeIcon: string
  options: CascaderOption[]
  value: string[]
  title: string
  optionKey: optionKey
  format: Record<string, string | number | null>
  closeable: boolean
  closeIconPosition: string
  closeIcon: string
  lazy: boolean
  onLoad: (node: any, resolve: any) => void
  onClose?: () => void
  onChange: (value: any, params: any) => void
  onPathChange: (value: any, params: any) => void
}

const defaultProps = {
  ...ComponentDefaults,
  activeColor: '',
  activeIcon: 'checklist',
  popup: true,
  visible: false,
  options: [],
  value: [],
  title: '',
  optionKey: { textKey: 'text', valueKey: 'value', childrenKey: 'children' },
  format: {},
  closeable: false,
  closeIconPosition: 'top-right',
  closeIcon: 'close',
  lazy: false,
  onLoad: () => {},
  onClose: () => {},
  onChange: () => {},
  onPathChange: () => {},
  ...Popup.defaultProps,
} as CascaderProps
const InternalCascader: ForwardRefRenderFunction<
  unknown,
  PropsWithChildren<Partial<CascaderProps>>
> = (props, ref) => {
  const {
    className,
    style,
    popup,
    visible,
    options,
    value,
    title,
    optionKey,
    format,
    closeable,
    closeIconPosition,
    closeIcon,
    lazy,
    onLoad,
    onClose,
    onChange,
    onPathChange,
  } = { ...defaultProps, ...props }

  const [tabvalue, setTabvalue] = useState('c1')
  const [optionsData, setOptionsData] = useState<CascaderPane[]>([])
  const isLazy = () => state.configs.lazy && Boolean(state.configs.onLoad)

  const [state] = useState({
    optionsData: [] as any,
    panes: [
      {
        nodes: [] as any,
        selectedNode: [] as CascaderOption | null,
        paneKey: '',
      },
    ],
    innerValue: value as CascaderValue,
    tree: new Tree([], {}),
    tabsCursor: 0, // 选中的tab项
    initLoading: false,
    currentProcessNode: [] as CascaderOption | null,
    configs: {
      lazy,
      onLoad,
      optionKey,
      format,
    },
    lazyLoadMap: new Map(),
  })

  const classPrefix = classNames(`nut-cascader`)
  const classesPane = classNames({
    [`${classPrefix}-pane`]: true,
  })

  useEffect(() => {
    initData()
  }, [options])

  useEffect(() => {
    if (value !== state.innerValue) {
      state.innerValue = value as CascaderValue
    }
  }, [value])

  const initData = async () => {
    // 初始化开始处理数据
    state.lazyLoadMap.clear()
    if (format && Object.keys(format).length > 0) {
      state.optionsData = convertListToOptions(
        options as CascaderOption[],
        format as format
      )
    } else {
      state.optionsData = options
    }

    state.tree = new Tree(state.optionsData as CascaderOption[], {
      value: state.configs.optionKey.valueKey,
      text: state.configs.optionKey.textKey,
      children: state.configs.optionKey.childrenKey,
    })
    if (isLazy() && !state.tree.nodes.length) {
      await invokeLazyLoad({
        root: true,
        loading: true,
        text: '',
        value: '',
      })
    }
    state.panes = [
      {
        nodes: state.tree.nodes,
        selectedNode: null,
        paneKey: 'c1',
      },
    ]
    syncValue()
    setOptionsData(state.panes)
  }
  // 处理有默认值时的数据
  const syncValue = async () => {
    const currentValue = state.innerValue
    if (currentValue === undefined || !state.tree.nodes.length) {
      return
    }

    if (currentValue.length === 0) {
      state.tabsCursor = 0
      return
    }

    let needToSync = currentValue

    if (isLazy() && Array.isArray(currentValue) && currentValue.length) {
      needToSync = []
      const parent: any = state.tree.nodes.find(
        (node) => node.value === currentValue[0]
      )

      if (parent) {
        needToSync = [parent.value]
        state.initLoading = true

        const last = await currentValue
          .slice(1)
          .reduce(async (p: Promise<CascaderOption | void>, value) => {
            const parent = await p
            await invokeLazyLoad(parent)
            const node: any = parent?.children?.find(
              (item: any) => item.value === value
            )
            if (node) {
              needToSync.push(value)
            }
            return Promise.resolve(node)
          }, Promise.resolve(parent))
        await invokeLazyLoad(last)
        state.initLoading = false
      }
    }

    if (needToSync.length && currentValue === value) {
      const pathNodes = state.tree.getPathNodesByValue(needToSync)
      pathNodes.forEach((node, index) => {
        state.tabsCursor = index
        // 当有默认值时，不触发 chooseItem 里的 emit 事件
        chooseItem(node, true)
      })
    }
  }

  const invokeLazyLoad = async (node?: CascaderOption | void) => {
    if (!node) {
      return
    }

    if (!state.configs.onLoad) {
      node.leaf = true
      return
    }

    if (
      state.tree.isLeaf(node, isLazy()) ||
      state.tree.hasChildren(node, isLazy())
    ) {
      return
    }

    node.loading = true

    const parent = node.root ? null : node
    let lazyLoadPromise = state.lazyLoadMap.get(node)

    if (!lazyLoadPromise) {
      lazyLoadPromise = new Promise((resolve) => {
        // 外部必须resolve
        state.configs.onLoad?.(node, resolve)
      })
      state.lazyLoadMap.set(node, lazyLoadPromise)
    }

    const nodes: CascaderOption[] | void = await lazyLoadPromise

    if (Array.isArray(nodes) && nodes.length > 0) {
      state.tree.updateChildren(nodes, parent)
    } else {
      // 如果加载完成后没有提供子节点，作为叶子节点处理
      node.leaf = true
    }
    node.loading = false
    state.lazyLoadMap.delete(node)
  }

  const close = () => {
    onClose && onClose()
  }

  const closePopup = () => {
    close()
  }

  /* type: 是否是静默模式，是的话不触发事件
  tabsCursor: tab的索引 */
  const chooseItem = async (node: CascaderOption, type: boolean) => {
    if ((!type && node.disabled) || !state.panes[state.tabsCursor]) {
      return
    }
    // 如果没有子节点
    if (state.tree.isLeaf(node, isLazy())) {
      node.leaf = true
      state.panes[state.tabsCursor].selectedNode = node
      state.panes = state.panes.slice(0, (node.level as number) + 1)
      if (!type) {
        const pathNodes = state.panes.map((item) => item.selectedNode)
        const optionParams = pathNodes.map((item: any) => item.value)
        onChange(optionParams, pathNodes)
        onPathChange(optionParams, pathNodes)
      }
      setOptionsData(state.panes)
      close()
      return
    }
    // 如果有子节点，滑到下一个
    if (state.tree.hasChildren(node, isLazy())) {
      const level = (node.level as number) + 1

      state.panes[state.tabsCursor].selectedNode = node
      state.panes = state.panes.slice(0, level)
      state.tabsCursor = level
      state.panes.push({
        nodes: node.children || [],
        selectedNode: null,
        paneKey: `c${state.tabsCursor + 1}`,
      })
      setOptionsData(state.panes)
      setTabvalue(`c${state.tabsCursor + 1}`)

      if (!type) {
        const pathNodes = state.panes.map((item) => item.selectedNode)
        const optionParams = pathNodes.map((item: any) => item?.value)
        onPathChange(optionParams, pathNodes)
      }
      return
    }
    state.currentProcessNode = node
    if (node.loading) {
      return
    }

    await invokeLazyLoad(node)
    if (state.currentProcessNode === node) {
      state.panes[state.tabsCursor].selectedNode = node
      chooseItem(node, type)
    }
    setOptionsData(state.panes)
  }

  const renderItem = () => {
    return (
      <div className={`${classPrefix} ${className}`} style={style}>
        {popup && <div className="nut-cascader__title">{title}</div>}
        <Tabs
          value={tabvalue}
          title={() => {
            return optionsData.map((pane, index) => (
              <div
                onClick={() => {
                  setTabvalue(pane.paneKey)
                  state.tabsCursor = index
                }}
                className={`nut-tabs__titles-item ${
                  tabvalue === pane.paneKey
                    ? 'nut-tabs__titles-item--active'
                    : ''
                }`}
                key={pane.paneKey}
              >
                <span className="nut-tabs__titles-item__text">
                  {!state.initLoading &&
                    state.panes.length &&
                    pane?.selectedNode?.text}
                  {!state.initLoading &&
                    state.panes.length &&
                    !pane?.selectedNode?.text &&
                    '请选择'}
                  {!(!state.initLoading && state.panes.length) && 'Loading...'}
                </span>
                <span className="nut-tabs__titles-item__line" />
              </div>
            ))
          }}
        >
          {!state.initLoading && state.panes.length ? (
            optionsData.map((pane) => (
              <Tabs.TabPane key={pane.paneKey} value={pane.paneKey}>
                <div className={classesPane}>
                  {pane.nodes &&
                    pane.nodes.map((node: any, index: number) => (
                      <CascaderItem
                        key={index}
                        {...props}
                        data={node}
                        checked={pane.selectedNode?.value === node.value}
                        chooseItem={(node: any) => chooseItem(node, false)}
                      />
                    ))}
                </div>
              </Tabs.TabPane>
            ))
          ) : (
            <Tabs.TabPane>
              <div className={classesPane} />
            </Tabs.TabPane>
          )}
        </Tabs>
      </div>
    )
  }

  return (
    <>
      {popup ? (
        <Popup
          // className="nut-cascadar-popup"
          visible={visible}
          position="bottom"
          round
          closeIcon={closeIcon}
          closeable={closeable}
          closeIconPosition={closeIconPosition}
          // todo 只关闭，不处理逻辑。和popup的逻辑不一致。关闭时需要增加是否要处理回调
          onClickOverlay={closePopup}
          onClickCloseIcon={closePopup}
        >
          {renderItem()}
        </Popup>
      ) : (
        renderItem()
      )}
    </>
  )
}

export const Cascader = React.forwardRef(InternalCascader)

Cascader.defaultProps = defaultProps
Cascader.displayName = 'NutCascader'
