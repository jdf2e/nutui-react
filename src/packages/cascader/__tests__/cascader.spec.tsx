import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Cascader } from '../cascader'

import { CascaderOption } from '../types'
import Tree from '../tree'
import { formatTree, convertListToOptions } from '../helper'

const later = (t = 0) => new Promise((r) => setTimeout(r, t))
const mockOptions = [
  {
    value: '浙江',
    text: '浙江',
    children: [
      {
        value: '杭州',
        text: '杭州',
        disabled: true,
        children: [
          { value: '西湖区', text: '西湖区' },
          { value: '余杭区', text: '余杭区' },
        ],
      },
      {
        value: '温州',
        text: '温州',
        children: [
          { value: '鹿城区', text: '鹿城区' },
          { value: '瓯海区', text: '瓯海区' },
        ],
      },
    ],
  },
  {
    value: '湖南',
    text: '湖南',
    disabled: true,
  },
  {
    value: '福建',
    text: '福建',
    children: [
      {
        value: '福州',
        text: '福州',
        children: [
          { value: '鼓楼区', text: '鼓楼区' },
          { value: '台江区', text: '台江区' },
        ],
      },
    ],
  },
]
const mockKeyConfigOptions = [
  {
    name: '浙江',
    items: [
      {
        name: '杭州',
        disabled: true,
        items: [{ name: '西湖区' }, { name: '余杭区' }],
      },
      {
        name: '温州',
        items: [{ name: '鹿城区' }, { name: '瓯海区' }],
      },
    ],
  },
  {
    name: '湖南',
    disabled: true,
  },
  {
    name: '福建',
    items: [
      {
        name: '福州',
        items: [{ name: '鼓楼区' }, { name: '台江区' }],
      },
    ],
  },
]
const mockConvertOptions = [
  { value: '北京', text: '北京', nodeId: 1, nodePid: 0, sort: 2 },
  { value: '朝阳区', text: '朝阳区', nodeId: 11, nodePid: 1 },
  { value: '亦庄', text: '亦庄', nodeId: 111, nodePid: 11 },
  { value: '广东省', text: '广东省', nodeId: 2, nodePid: 0, sort: 1 },
  { value: '广州市', text: '广州市', nodeId: 21, nodePid: 2 },
]

describe('helpers', () => {
  test('formatTree', () => {
    const fromatedTree = formatTree(mockKeyConfigOptions, null, {
      children: 'items',
      text: 'name',
      value: 'name',
    })

    expect(fromatedTree).toMatchObject(mockOptions)
  })

  test('convertListToOptions', () => {
    const convertList = convertListToOptions(mockConvertOptions, {
      topId: 0,
      idKey: 'nodeId',
      pidKey: 'nodePid',
      sortKey: '',
    })
    expect(convertList).toMatchObject([
      {
        nodePid: 0,
        nodeId: 1,
        text: '北京',
        value: '北京',
        sort: 2,
        children: [
          {
            nodePid: 1,
            nodeId: 11,
            text: '朝阳区',
            value: '朝阳区',
            children: [
              {
                nodePid: 11,
                nodeId: 111,
                text: '亦庄',
                value: '亦庄',
                children: [],
              },
            ],
          },
        ],
      },
      {
        nodePid: 0,
        nodeId: 2,
        text: '广东省',
        value: '广东省',
        children: [
          {
            nodePid: 2,
            nodeId: 21,
            text: '广州市',
            value: '广州市',
          },
        ],
      },
    ])
  })
})

describe('Tree', () => {
  test('tree', () => {
    const tree = new Tree([
      {
        text: '浙江',
        value: '浙江',
      },
      {
        text: '福建',
        value: '福建',
      },
    ])
    expect(tree.nodes).toMatchObject([
      {
        text: '浙江',
        value: '浙江',
      },
      {
        text: '福建',
        value: '福建',
      },
    ])
  })

  test('tree with config', () => {
    const tree = new Tree(mockKeyConfigOptions, {
      value: 'name',
      text: 'name',
      children: 'items',
    })
    expect(tree.nodes).toMatchObject(mockOptions)
  })

  const tree = new Tree(mockOptions)
  test('getPathNodesByValue', () => {
    const pathNodes = tree.getPathNodesByValue(['浙江', '杭州', '西湖区'])
    const mappedPathNodes = pathNodes.map(({ text, value }) => ({
      text,
      value,
    }))

    expect(mappedPathNodes).toMatchObject([
      { text: '浙江', value: '浙江' },
      { text: '杭州', value: '杭州' },
      { text: '西湖区', value: '西湖区' },
    ])
  })

  test('isLeaf', () => {
    const node = tree.getNodeByValue('西湖区')
    let isLeaf = tree.isLeaf(node as CascaderOption, false)
    expect(isLeaf).toBeTruthy()
    isLeaf = tree.isLeaf(node as CascaderOption, true)
    expect(isLeaf).toBeFalsy()
  })

  test('hasChildren', () => {
    let node = tree.getNodeByValue('西湖区')

    let hasChildren = tree.hasChildren(node as CascaderOption, false)
    expect(hasChildren).toBeFalsy()

    hasChildren = tree.hasChildren(node as CascaderOption, true)
    expect(hasChildren).toBeFalsy()

    node = tree.getNodeByValue('杭州')

    hasChildren = tree.hasChildren(node as CascaderOption, false)
    expect(hasChildren).toBeTruthy()

    hasChildren = tree.hasChildren(node as CascaderOption, true)
    expect(hasChildren).toBeTruthy()
  })

  test('updateChildren', () => {
    let node = tree.getNodeByValue('福建')
    expect(node).toBeTruthy()

    tree.updateChildren(
      [{ text: '福州', value: '福州' }],
      node as CascaderOption
    )
    node = tree.getNodeByValue('福州') as CascaderOption
    expect(node).toBeTruthy()
    expect(node.value).toBe('福州')

    tree.updateChildren(
      [{ text: '鼓楼区', value: '鼓楼区' }],
      node as CascaderOption
    )
    node = tree.getNodeByValue('鼓楼区') as CascaderOption
    expect(node).toBeTruthy()
    expect(node.value).toBe('鼓楼区')
  })

  // test('updateChildren with CascaderConfig', () => {
  //   const tree = new Tree(
  //     [
  //       {
  //         name: '福建',
  //         items: [{ name: '福州' }],
  //       },
  //     ],
  //     {
  //       value: 'name',
  //       text: 'name',
  //       children: 'items',
  //     }
  //   )
  //   expect(tree.nodes).toMatchObject([
  //     {
  //       text: '福建',
  //       value: '福建',
  //       children: [{ text: '福州', value: '福州' }],
  //     },
  //   ])

  //   let node = tree.getNodeByValue('福州') as CascaderOption
  //   expect(node).toBeTruthy()
  //   tree.updateChildren([{ name: '鼓楼区' }], node)
  //   node = tree.getNodeByValue('鼓楼区') as CascaderOption
  //   expect(node).toBeTruthy()
  //   expect(node).toMatchObject({
  //     text: '鼓楼区',
  //     value: '鼓楼区',
  //   })
  // })
})

describe('Cascader', () => {
  it('options', async () => {
    const { container } = render(
      <Cascader value={['福建', '福州', '鼓楼区']} options={mockOptions} />
    )
    expect(container).toMatchSnapshot()
  })

  it('options with valueKey/textKey/childrenKey', async () => {
    const { container } = render(
      <Cascader
        value={['福建', '福州', '鼓楼区']}
        options={mockKeyConfigOptions}
        optionKey={{
          valueKey: 'name',
          textKey: 'name',
          childrenKey: 'items',
        }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('options with format', async () => {
    const { container } = render(
      <Cascader
        value={['广东省', '广州市']}
        options={mockConvertOptions}
        format={{
          topId: 0,
          idKey: 'nodeId',
          pidKey: 'nodePid',
          sortKey: 'sort',
        }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('visible false', async () => {
    const { container } = render(
      <Cascader visible={false} options={mockOptions} />
    )
    expect(container.querySelector('.nut-popup')).toBeNull()
  })

  it('visible true', async () => {
    const { container } = render(
      <Cascader
        visible
        value={['福建', '福州', '鼓楼区']}
        options={mockOptions}
      />
    )
    expect(container.querySelector('.nut-popup')).toBe
    expect(container).toMatchSnapshot()
    // 点击叶子节点时关闭popup
    const el = container.querySelectorAll('.nut-cascader-pane')[2].childNodes[0]
    fireEvent.click(el)
    // expect(
    //   container.querySelectorAll('.nut-cascader-pane')[2].childNodes[0]
    // ).toEqual(
    //   '<div class="active nut-cascader-item"><div class="nut-cascader-item__title">鼓楼区</div><i class="nut-cascader-item__icon-check nut-icon nutui-iconfont nut-icon-checklist"/></div>'
    // )
  })

  it('value', async () => {
    const { container } = render(<Cascader options={mockOptions} />)
    expect(
      container.querySelectorAll('.nut-cascader-item[aria-checked="true"]')
        .length
    ).toBe(0)
    expect(container).toMatchSnapshot()
  })

  it('select', async () => {
    const change = jest.fn()
    const pathChange = jest.fn()
    const { container } = render(
      <Cascader
        visible
        value={['福建', '福州', '鼓楼区']}
        options={mockOptions}
        onChange={change}
        onPathChange={pathChange}
      />
    )
    // 模拟点击
    const pane = container.querySelectorAll('.nut-cascader-pane')[2]
    fireEvent.click(pane)
    const item = pane.childNodes[0]
    // console.log('item', item)
    fireEvent.click(item)
    // let pathChange: any = container.emitted().pathChange[0];
    expect(pathChange).toBeCalled()
    // ...
  })

  // it('value with lazy', async () => {
  //   const { container } = render(
  //     <Cascader
  //       visible
  //       value={['A0', 'A12', 'A21']}
  //       lazy
  //       onLoad={(node: any, resolve: (children: any) => void) => {
  //         setTimeout(() => {
  //           if (node.root) {
  //             resolve([
  //               { value: 'A0', text: 'A0' },
  //               { value: 'B0', text: 'B0' },
  //               { value: 'C0', text: 'C0' },
  //             ])
  //           } else {
  //             const { value, level } = node
  //             const text = value.substring(0, 1)
  //             const value1 = `${text}${level + 1}1`
  //             const value2 = `${text}${level + 1}2`
  //             resolve([
  //               { value: value1, text: value1, leaf: level >= 1 },
  //               { value: value2, text: value2, leaf: level >= 1 },
  //             ])
  //           }
  //         }, 50)
  //       }}
  //     />
  //   )
  //   await later(160)
  //   expect(container).toMatchSnapshot()
  // })

  it('select with lazy', async () => {
    const { container } = render(
      <Cascader
        lazy
        onLoad={(node: any, resolve: (children: any) => void) => {
          setTimeout(() => {
            setTimeout(() => {
              // root表示第一层数据
              if (node.root) {
                resolve([
                  { value: 'A0', text: 'A0' },
                  { value: 'B0', text: 'B0' },
                ])
              } else {
                const { value, level } = node
                const text = value.substring(0, 1)
                const value1 = `${text}${level + 1}1`
                const value2 = `${text}${level + 1}2`
                resolve([
                  { value: value1, text: value1, leaf: level >= 1 },
                  { value: value2, text: value2, leaf: level >= 1 },
                ])
              }
            }, 50)
          }, 50)
        }}
      />
    )

    expect(container).toMatchSnapshot()
    await later(160)
    expect(container).toMatchSnapshot()
    // ...
  })

  it('change tab', async () => {
    const change = jest.fn()
    const pathChange = jest.fn()
    const { container } = render(
      <Cascader
        visible
        value={['福建', '福州', '鼓楼区']}
        options={mockOptions}
        onChange={change}
        onPathChange={pathChange}
      />
    )

    expect(container).toMatchSnapshot()

    expect(container.querySelector('.nut-popup')).toBe

    const tabPane = container.querySelectorAll('.nut-tabs__titles-item')[0]
    fireEvent.click(tabPane)
    expect(container).toMatchSnapshot()
  })
})
