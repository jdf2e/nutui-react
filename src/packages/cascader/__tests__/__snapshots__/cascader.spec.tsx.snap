// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cascader change tab 1`] = `
<div>
  <div
    class="nut-overlay"
    style="--nutui-overlay-zIndex: 1000;"
  />
  <div
    class="nut-popup-round nut-popup nut-popup-bottom "
    style="z-index: 1000;"
  >
    <div
      class="nut-cascader "
    >
      <div
        class="nut-tabs nut-tabs--horizontal"
        value="c1"
      >
        <div
          class="nut-tabs__titles nut-tabs__titles--line nut-tabs__titles--scrollable"
        >
          <div
            class="nut-tabs__titles-item nut-tabs__titles-item--active"
          >
            <span
              class="nut-tabs__titles-item__text"
            >
              请选择
            </span>
            <span
              class="nut-tabs__titles-item__line"
            />
          </div>
        </div>
        <div
          class="nut-tabs__content__wrap"
        >
          <div
            class="nut-tabs__content"
            style="transform: translate3d(--100%, 0, 0); transition-duration: 300ms;"
          >
            <div
              class="active nut-tabpane"
            >
              <div
                class="nut-cascader-pane"
              >
                <div
                  class="nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    浙江
                  </div>
                </div>
                <div
                  class="disabled nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    湖南
                  </div>
                </div>
                <div
                  class="nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    福建
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Cascader change tab 2`] = `
<div>
  <div
    class="nut-overlay"
    style="--nutui-overlay-zIndex: 1000;"
  />
  <div
    class="nut-popup-round nut-popup nut-popup-bottom "
    style="z-index: 1000;"
  >
    <div
      class="nut-cascader "
    >
      <div
        class="nut-tabs nut-tabs--horizontal"
        value="c1"
      >
        <div
          class="nut-tabs__titles nut-tabs__titles--line nut-tabs__titles--scrollable"
        >
          <div
            class="nut-tabs__titles-item nut-tabs__titles-item--active"
          >
            <span
              class="nut-tabs__titles-item__text"
            >
              请选择
            </span>
            <span
              class="nut-tabs__titles-item__line"
            />
          </div>
        </div>
        <div
          class="nut-tabs__content__wrap"
        >
          <div
            class="nut-tabs__content"
            style="transform: translate3d(--100%, 0, 0); transition-duration: 300ms;"
          >
            <div
              class="active nut-tabpane"
            >
              <div
                class="nut-cascader-pane"
              >
                <div
                  class="nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    浙江
                  </div>
                </div>
                <div
                  class="disabled nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    湖南
                  </div>
                </div>
                <div
                  class="nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    福建
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Cascader options 1`] = `<div />`;

exports[`Cascader options with format 1`] = `<div />`;

exports[`Cascader options with valueKey/textKey/childrenKey 1`] = `<div />`;

exports[`Cascader value 1`] = `<div />`;

exports[`Cascader visible true 1`] = `
<div>
  <div
    class="nut-overlay"
    style="--nutui-overlay-zIndex: 1000;"
  />
  <div
    class="nut-popup-round nut-popup nut-popup-bottom "
    style="z-index: 1000;"
  >
    <div
      class="nut-cascader "
    >
      <div
        class="nut-tabs nut-tabs--horizontal"
        value="c3"
      >
        <div
          class="nut-tabs__titles nut-tabs__titles--line nut-tabs__titles--scrollable"
        >
          <div
            class="nut-tabs__titles-item "
          >
            <span
              class="nut-tabs__titles-item__text"
            >
              福建
            </span>
            <span
              class="nut-tabs__titles-item__line"
            />
          </div>
          <div
            class="nut-tabs__titles-item "
          >
            <span
              class="nut-tabs__titles-item__text"
            >
              福州
            </span>
            <span
              class="nut-tabs__titles-item__line"
            />
          </div>
          <div
            class="nut-tabs__titles-item nut-tabs__titles-item--active"
          >
            <span
              class="nut-tabs__titles-item__text"
            >
              鼓楼区
            </span>
            <span
              class="nut-tabs__titles-item__line"
            />
          </div>
        </div>
        <div
          class="nut-tabs__content__wrap"
        >
          <div
            class="nut-tabs__content"
            style="transform: translate3d(-200%, 0, 0); transition-duration: 300ms;"
          >
            <div
              class="nut-tabpane"
            >
              <div
                class="nut-cascader-pane"
              >
                <div
                  class="nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    浙江
                  </div>
                </div>
                <div
                  class="disabled nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    湖南
                  </div>
                </div>
                <div
                  class="active nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    福建
                  </div>
                  <svg
                    aria-labelledby="Checklist"
                    class="nut-cascader__icon-check nut-icon nut-icon-Checklist"
                    role="presentation"
                    viewBox="0 0 1024 1024"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M159.289 500.622c62.578 0 125.155 17.067 221.867 102.4 8.533 5.69 19.91 5.69 25.6 0 48.355-54.044 238.933-261.689 455.11-329.955 0 0 28.445-5.69 42.667 19.91 8.534 17.067 19.911 34.134-5.689 54.045-22.755 17.067-264.533 179.2-440.888 440.89l-2.845 2.844c-11.378 8.533-68.267 51.2-119.467-14.223-56.888-71.11-85.333-139.377-196.266-196.266-2.845 0-2.845-2.845-5.69-5.69-11.377-11.377-54.044-73.955 25.6-73.955z"
                      fill="currentColor"
                      fill-opacity="0.9"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="nut-tabpane"
            >
              <div
                class="nut-cascader-pane"
              >
                <div
                  class="active nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    福州
                  </div>
                  <svg
                    aria-labelledby="Checklist"
                    class="nut-cascader__icon-check nut-icon nut-icon-Checklist"
                    role="presentation"
                    viewBox="0 0 1024 1024"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M159.289 500.622c62.578 0 125.155 17.067 221.867 102.4 8.533 5.69 19.91 5.69 25.6 0 48.355-54.044 238.933-261.689 455.11-329.955 0 0 28.445-5.69 42.667 19.91 8.534 17.067 19.911 34.134-5.689 54.045-22.755 17.067-264.533 179.2-440.888 440.89l-2.845 2.844c-11.378 8.533-68.267 51.2-119.467-14.223-56.888-71.11-85.333-139.377-196.266-196.266-2.845 0-2.845-2.845-5.69-5.69-11.377-11.377-54.044-73.955 25.6-73.955z"
                      fill="currentColor"
                      fill-opacity="0.9"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="active nut-tabpane"
            >
              <div
                class="nut-cascader-pane"
              >
                <div
                  class="active nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    鼓楼区
                  </div>
                  <svg
                    aria-labelledby="Checklist"
                    class="nut-cascader__icon-check nut-icon nut-icon-Checklist"
                    role="presentation"
                    viewBox="0 0 1024 1024"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M159.289 500.622c62.578 0 125.155 17.067 221.867 102.4 8.533 5.69 19.91 5.69 25.6 0 48.355-54.044 238.933-261.689 455.11-329.955 0 0 28.445-5.69 42.667 19.91 8.534 17.067 19.911 34.134-5.689 54.045-22.755 17.067-264.533 179.2-440.888 440.89l-2.845 2.844c-11.378 8.533-68.267 51.2-119.467-14.223-56.888-71.11-85.333-139.377-196.266-196.266-2.845 0-2.845-2.845-5.69-5.69-11.377-11.377-54.044-73.955 25.6-73.955z"
                      fill="currentColor"
                      fill-opacity="0.9"
                    />
                  </svg>
                </div>
                <div
                  class="nut-cascader-item"
                >
                  <div
                    class="nut-cascader-item__title"
                  >
                    台江区
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Cascader visible true 2`] = `
<div
  class="active nut-cascader-item"
>
  <div
    class="nut-cascader-item__title"
  >
    鼓楼区
  </div>
  <svg
    aria-labelledby="Checklist"
    class="nut-cascader__icon-check nut-icon nut-icon-Checklist"
    role="presentation"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M159.289 500.622c62.578 0 125.155 17.067 221.867 102.4 8.533 5.69 19.91 5.69 25.6 0 48.355-54.044 238.933-261.689 455.11-329.955 0 0 28.445-5.69 42.667 19.91 8.534 17.067 19.911 34.134-5.689 54.045-22.755 17.067-264.533 179.2-440.888 440.89l-2.845 2.844c-11.378 8.533-68.267 51.2-119.467-14.223-56.888-71.11-85.333-139.377-196.266-196.266-2.845 0-2.845-2.845-5.69-5.69-11.377-11.377-54.044-73.955 25.6-73.955z"
      fill="currentColor"
      fill-opacity="0.9"
    />
  </svg>
</div>
`;
